astropylibrarian/keywords.py

```diff
@@ -0,0 +1,146 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""Standardized Learn Astropy keywords."""
+
+from pathlib import Path
+from typing import Dict, List, Optional
+
+import yaml
+
+KeywordTable = Dict[str, List[str]]
+"""Keyword table data type.
+
+The canonical keyword is the key, and alternative forms are strings in the
+list.
+"""
+
+
+class KeywordDb:
+    """A database of Astropy keywords and sorter of keyword types.
+
+    Parameters
+    ----------
+    astropy_package : `KeywordTable`
+        Keywords in the "Astropy package" group.
+    python_package : `KeywordTable`
+        Keywords in the "Python package" group.
+    task : `KeywordTable`
+        Keywords in the "task" group.
+    science : `KeywordTable`
+        Keywords in the "science" group.
+    """
+
+    def __init__(
+        self, *, astropy_package: KeywordTable,
+        python_package: KeywordTable, task: KeywordTable,
+        science: KeywordTable
+    ) -> None:
+        self._astropy_package_keywords = astropy_package
+        self._python_package_keywords = python_package
+        self._task_keywords = task
+        self._science_keywords = science
+
+    @classmethod
+    def load(cls, path: Optional[Path] = None) -> 'KeywordDb':
+        """Load a KeywordDB from a YAML file.
+
+        Parameters
+        ----------
+        path : `pathlib.Path`, optional
+            Path to the YAML-formatted keyword database file. Leave as `None`
+            to use the built-in keyword database.
+
+        Returns
+        -------
+        KeywordDb
+            A keyword database instance.
+        """
+        if path is None:
+            path = Path(__file__).parent / 'data' / 'keywords.yaml'
+
+        db = yaml.safe_load(path.read_text())
+
+        astropy_package_keywords = cls._load_keyword_table(
+            db['astropy_package'])
+        python_package_keywords = cls._load_keyword_table(
+            db['python_package'])
+        task_keywords = cls._load_keyword_table(
+            db['task'])
+        science_keywords = cls._load_keyword_table(
+            db['science'])
+
+        return cls(
+            astropy_package=astropy_package_keywords,
+            python_package=python_package_keywords,
+            task=task_keywords,
+            science=science_keywords
+        )
```

Comment:
This makes me think that we could also let the yaml file dictate the keyword groups. For example, sometime like:
```python
kw_groups = {}
for kw_group_name in  db.keys():
    kw_groups[kw_group_name] = cls._load_keyword_table(db[kw_group_name])

return cls(**kw_groups)
```

And then the initializer would need to instead accept `**kwargs`, and the `astropy_package_keywords` etc. properties on TutorialRecord would either have to become a method `get_package_keyword(keyword_name)`, or do some trickery with `__getattr__` to autopopulate the property names...

Anyway, just a thought - imagining the possibility of wanting to change or add keyword group names and thinking about all the places we would need to change things. (this would also reduce some code duplication!)

Reply:
Nice, I like that idea. I've implemented in the revised KeywordsDB.